Add Bucket Storage Support for files

masudur-rahman opened the pull request:
- Add bucket storage support for the following items:
	- `avatars`
	- `repo-avatars`
	- `attachments`
	- `lfs`

lafriks commented on modules/setting/setting.go:
```diff
@@ -613,7 +621,7 @@ func NewContext() {
 	DisableRouterLog = sec.Key("DISABLE_ROUTER_LOG").MustBool()
 	StaticRootPath = sec.Key("STATIC_ROOT_PATH").MustString(AppWorkPath)
-	AppDataPath = sec.Key("APP_DATA_PATH").MustString(path.Join(AppWorkPath, "data"))
+	AppDataPath = sec.Key("APP_DATA_PATH").MustString("data")
```
This should not be changed

masudur-rahman replied:
I need the followings:
```go
AvatarUploadPath = "data/avatars/"
LFS.ContentPath	 = "data/lfs/"
AttachmentPath	 = "data/attachments/"
RepositoryAvatarUploadPath = "data/repo-avatars/"
```

Should I use:
`LFS.ContentPath = sec.Key("LFS_CONTENT_PATH").MustString(filepath.Join("data", "lfs/"))`
instead of:
`LFS.ContentPath = sec.Key("LFS_CONTENT_PATH").MustString(filepath.Join(AppDataPath, "lfs"))`

lafriks replied:
Problem is that this is just defaults. User could have specified completely different paths in their app.ini for ex:
```ini
AvatarUploadPath = "/srv/avatars/upload/"
LFS.ContentPath	 = "/srv/lfs/"
AttachmentPath	 = "/srv/attachments/"
RepositoryAvatarUploadPath = "/srv/avatars/repo-upload/"
```

And in such configuration your code will not work for local storage anymore

masudur-rahman replied:
It will work for either local or any other cloud storage.
By default I set it to local storage. https://github.com/go-gitea/gitea/blob/16ac0bb98f889f15628e1779889fd0b689927f2a/modules/setting/file_storage.go#L10-L23
If user wants to use any other cloud storage or other local storage, they just have to specify it in their `app.ini`

Users are free to provide their own upload path like you suggested and it will work fine.

lafriks replied:
I meant it should work for local storage same as it was before without changing app.ini

masudur-rahman replied:
Yes, it will work as it was before. Just the files will be stored in the directory specified in `FileStorage.BucketURL`.

So, what do you suggest about https://github.com/go-gitea/gitea/pull/7795#discussion_r311939134 ?

guillep2k replied:
So I've checked out this PR and ran it with my usual test settings:

```
[server]
SSH_DOMAIN       = ....
DOMAIN           =  ....
HTTP_PORT        =  ....
PROTOCOL         = http
ROOT_URL         =  ....
OFFLINE_MODE     = false
APP_DATA_PATH    = /mnt/external/my_test_data
```

Then I've changed my avatar. I _expected_ the avatar to be created at:
```
/mnt/external/my_test_data/avatars/a9050bf53d1005dfe5398ce9fa5b07aa
```

Instead, with this PR, the avatar was **incorrectly** created at:
```
/home/giteatest/avatars/a9050bf53d1005dfe5398ce9fa5b07aa
```

As you can see, with this PR Gitea completely ignored the value of my `APP_DATA_PATH`. This is the breaking change we've been talking about. The code must behave **like it did before this PR** if the user doesn't change their `app.ini`. Users that upgrade to this must not notice any difference, unless they actively decide to use the new bucket storage capability.

I hope that this example helps.

masudur-rahman replied:
Hey @guillep2k, thanks for taking the time for testing.

If `AVATAR_UPLOAD_PATH` is set either relative or absolute, it ignores `APP_DATA_PATH` as it was ignored previously. https://github.com/go-gitea/gitea/blob/master/modules/setting/setting.go#L850-L854

But if `AVATAR_UPLOAD_PATH` is not specified, then it uses `APP_DATA_PATH`+"avatars" to read or store the avatars.

Either way if the `AvatarUploadPath` is relative then the whole path to store the avatars is `AppWorkPath`+`AvatarUploadPath`.
If the path is absolute the avatars is stored to `AvatarUploadPath`.
It works like it did previously.

I've attached the scenarios here : https://github.com/go-gitea/gitea/blob/e29ba594b0/modules/storage/storage.go#L124-L154

May be there was some errors in my previous commit but I refactored the code and I hope it works like it did before this pr.

**Please, pull again and check if it's now okay. Thanks again.**

lafriks commented on modules/setting/setting.go:
```diff
@@ -685,9 +693,8 @@ func NewContext() {
-		LFS.ContentPath = filepath.Join(AppWorkPath, LFS.ContentPath)
-	}
+	forcePathSeparator(LFS.ContentPath)
+	LFS.ContentPath = suffixPathSeparator(LFS.ContentPath)
```
This breaks current behavior and is not really related to this PR

masudur-rahman commented:
Hey @lafriks , previously I misunderstood you.
I have changed the code to be backward compatible. Please, let me know if anything needs to be changed.

lafriks commented on docs/content/doc/advanced/config-cheat-sheet.en-us.md:
```diff
@@ -507,8 +507,12 @@ Two special environment variables are passed to the render command:
 - `SHOW_FOOTER_VERSION`: **true**: Show Gitea version information in the footer.
-- `SHOW_FOOTER_TEMPLATE_LOAD_TIME`: **true**: Show time of template execution in the footer.
\ No newline at end of file
+- `SHOW_FOOTER_TEMPLATE_LOAD_TIME`: **true**: Show time of template execution in the footer.
```
This change does not seems to be related to this PR

masudur-rahman replied:
It's just a new line at the end.

lafriks commented on models/migrations/migrations.go:
```diff
@@ -638,14 +640,16 @@ func attachmentRefactor(x *xorm.Engine) error {
-			isSucceed = false
-			return err
-		}
+		if filepath.IsAbs(setting.AttachmentPath) {
```
Why checking only for absolute attachment path?

masudur-rahman replied:
```
AttachmentPath	 = "/srv/attachments/"
```
If the AttachmentPath is absolute, prefix "file://" must be added to it in order to open it using `blob.OpenBucket`.
That's why I checked it.

masudur-rahman replied:
I have updated it. I've attached a summary how whole `BucketURL` and `AttachmentPath` or others work.
https://github.com/go-gitea/gitea/blob/4c390c9699ee6400a9467c4c8234af9efba5acaf/modules/setting/setting.go#L1053-L1083

Please take a look if it seems okay. Thanks

lafriks commented on models/migrations/v61.go:
```diff
@@ -29,14 +32,35 @@ func addSizeToAttachment(x *xorm.Engine) error {
+	var bucket *blob.Bucket
+	var err error
+	ctx := context.Background()
+	if filepath.IsAbs(setting.AttachmentPath) {
```
IMHO it should be checking for some other setting not if path is absolute

lunny commented on docs/content/doc/advanced/config-cheat-sheet.en-us.md.orig:
```diff
@@ -0,0 +1,518 @@
+---
```
This file should not be added to git.

lunny commented on models/admin.go:
```diff
@@ -5,8 +5,13 @@
+	"path/filepath"
+
+	"code.gitea.io/gitea/modules/setting"
+	"gocloud.dev/blob"
```
Wrong imports order.

lunny commented:
I think we should have an abstract layout under `modules/` i.e. `modules/fs` to let people chose file system or block storage.

guillep2k commented on custom/conf/app.ini.sample:
```diff
@@ -792,3 +792,9 @@ IS_INPUT_FILE = false
+
+[storage]
+; URL of Bucket where files, such as, attachments, avatars, etc. are stored
+; If unset, file://<current_directory> is used to maintain backward compatibility
```
"`backward compatibility`" makes it sound like buckets are the "modern" way to go. Could this be changed to something like "`if unset, {APP_DATA_PATH}/attachments is used`" (not checked) or something around those lines?

masudur-rahman replied:
Actually if unset, it now uses `AppWorkPath`. When using this path for `attachments` or `avatars`, `AttachmentPath` or `AvatarUploadPath` is appended to this `BucketURL`.
If `AttachmentPath` is unset, we set this as `data/attachments`, to maintain similarity between all the available buckets. https://github.com/go-gitea/gitea/blob/83ccf4c649e58abb83c7d21ec42a3f378be51949/modules/setting/setting.go#L803

guillep2k replied:
So, if I understand well, the user can't use `APP_DATA_PATH` anymore? I think it's too big of a breaking change. If the user upgrades because they were interested in X other functionality the new version provides, he or she might not be aware of this change and get in a lot of trouble. Can't they both co-exist? e.g. if there are both `APP_DATA_PATH` and `BUCKET_URL`, then the former should be ignored, but if there's only `APP_DATA_PATH`, the old behaviour should be honored.

masudur-rahman replied:
`APP_DATA_PATH` related issue is now fixed. 
https://github.com/go-gitea/gitea/blob/4c390c9699ee6400a9467c4c8234af9efba5acaf/modules/setting/setting.go#L1053-L1083

Now, `APP_DATA_PATH`and `BUCKET_URL` both co-exist. Please let me know if I did it right.

guillep2k commented on models/admin.go:
```diff
@@ -60,6 +62,23 @@ func RemoveAllWithNotice(title, path string) {
+	ctx := context.Background()
+	bucket, err := setting.OpenBucket(ctx, bucketPath)
+	if err != nil {
+		return fmt.Errorf("could not open bucket: %v", err)
```
Is there a way to make an error here gracefully fail, so the user can attempt to retry the operation? Third-party requests are likely to fail at some point.

masudur-rahman replied:
The error related to `OpenBucket` is handled by the bucket providers using `exponential backoff` algo. I think it's okay not to handle it here.

guillep2k commented on models/attachment.go:
```diff
@@ -66,38 +67,56 @@ func (a *Attachment) LocalPath() string {
+	ctx := context.Background()
+	bucket, err := setting.OpenBucket(ctx, setting.AttachmentPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not open bucket: %v", err)
```
Same thing as above about gracefully return connection errors.

masudur-rahman commented:
> I think we should have an abstract layout under `modules/` i.e. `modules/fs` to let people chose file system or block storage.

Hey @lunny, I am working on that now. I will make those changes in a new pr. If the changes look good, we can copy the changes in this pr later. 
Please take a look if the changes are okay or any thing needs to be changed. Thanks

masudur-rahman commented:
@lunny , I have refactored this pr. I moved the `FileStorage` thing to `modules/storage`.
https://github.com/go-gitea/gitea/blob/e29ba594b069ae3011601ba210b40cbe9a057ff1/modules/storage/storage.go

I hope these changes serve the abstract layout you asked for. 
Please let me know if everything's okay. Thanks.

tamalsaha commented:
@lunny / @lafriks / @guillep2k and folks, if you can take another look at it that will be great.

guillep2k commented:
Now the `APP_DATA_PATH` setting seems to be working (at least for avatars), but for some reason uploading a picture leaves an `.attrs` file behind that I don't think I've seen before:

```
-rw-------+ 1 giteatest giteatest 269836 Aug 22 22:17 my_test_data/avatars/9-eae2c62356750433ffc3c52ffb9ceee9
-rw-rw----+ 1 giteatest giteatest    196 Aug 22 22:17 my_test_data/avatars/9-eae2c62356750433ffc3c52ffb9ceee9.attrs
```

Contents:
```
{"user.cache_control":"","user.content_disposition":"","user.content_encoding":"","user.content_language":"","user.content_type":"image/png","user.metadata":null,"md5":"a+USfqDVjIMq4Sy3HXOWdA=="}
```

Why is it there? Is that a side-effect of your changes?

tamalsaha commented:
@guillep2k , this is a result of using the gocloud/blob/filestorage api. This library stores file attributes in this .attrs file .

- https://github.com/google/go-cloud/blob/master/blob/fileblob/fileblob.go
- https://github.com/google/go-cloud/pull/203

tamalsaha commented:
Please take a look. The git conflicts are fixed.

guillep2k commented:
Weird fail in the build. Somebody will restart it soon.

lunny commented:
I will review this ASAP.

masudur-rahman commented:
@lunny, I keep getting `build fail` after resolving the conflicts. I ran `go mod tidy && go mod vendor` every time after resolving the conflicts. 
But there is some dissimilarity in `vendor/modules.txt` between my version and `continuous-integration/drone/pr` version which I don't know why.

Can you please give me some suggestions what should I do to resolve this problem ?

lunny commented:
@masudur-rahman CI failed because of other reason and just fixed. Please resolve the conflicts.

masudur-rahman commented:
@lunny, I resolved the conflicts and kept the changes in a single commit. Still getting `build fail` due to `modules.txt`.

masudur-rahman commented:
At last the `continuous-integration/drone/pr` is passing. May be it will be failing after merging some other pr. 
@lunny , @lafriks & @guillep2k , can you guys please take a look while `continuous-integration/drone/pr` is passing ?

Thanks...

lunny commented on models/attachment.go:
```diff
@@ -75,29 +81,32 @@ func (a *Attachment) DownloadURL() string {
 		return nil, fmt.Errorf("Create: %v", err)
 	}
-	defer fw.Close()
 
```
just keep `defer fw.Close()` is better.

masudur-rahman replied:
`defer fw.Close()` can't be used here.
`fw` must be closed before `fi, err := fs.Attributes()` (Line 105). It can't open the file for attributes unless the `fw.Close()` is called.

That's why I had to do like this.

masudur-rahman replied:
Sorry for the late. 
Your suggestion will work fine here and will do that.

guillep2k commented on models/repo.go:
```diff
@@ -2684,11 +2720,13 @@ func (repo *Repository) UploadAvatar(data []byte) error {
+	fs := storage.FileStorage{
+		Ctx:      context.Background(),
+		Path:     setting.RepositoryAvatarUploadPath,
+		FileName: repo.Avatar,
```
What happens if two users choose the same file name for their repo avatars?

masudur-rahman replied:
Here, `<repo.ID>-<md5.Sum(data)>` is used as  `repo.Avatar`. So even if two user uses same file,  repo.ID will differ them.
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/models/repo.go#L2718

guillep2k commented on models/user.go:
```diff
@@ -520,16 +530,18 @@ func (u *User) UploadAvatar(data []byte) error {
+	fs := storage.FileStorage{
+		Ctx:      context.Background(),
+		Path:     setting.AvatarUploadPath,
+		FileName: u.Avatar,
```
Same question as for repo avatar file name.

masudur-rahman replied:
Same answer as for repo.avatar.
But previously, if two users uploaded the same file, it would replace the previous one (also in previous system).
https://github.com/go-gitea/gitea/blob/d4cd4ed4422be222088dd8535a228b906f43fdc2/models/user.go#L523

I used same format as `repo.Avatar` to avoid this collision.
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/models/user.go#L533

guillep2k commented on modules/migrations/gitea.go:
```diff
@@ -217,12 +218,12 @@ func (g *GiteaLocalUploader) CreateReleases(releases ...*base.Release) error {
-			localPath := attach.LocalPath()
-			if err = os.MkdirAll(path.Dir(localPath), os.ModePerm); err != nil {
-				return fmt.Errorf("MkdirAll: %v", err)
+			fs := storage.FileStorage{
```
This is trickier. Depeding on the current version value you should or should not ignore the app.ini setting and use local file storage in this function.

guillep2k commented on modules/public/avatars.go:
```diff
@@ -0,0 +1,91 @@
+// AvatarOptions represents the available options to configure the AvatarHandler.
+type AvatarOptions struct {
+	Prefix      string
+	SkipLogging bool
```
Is this necessary? The new logger system allows to set logging levels for different features individually.

masudur-rahman replied:
I followed the existing `Options` structure to handle this part. 
https://github.com/go-gitea/gitea/blob/d4cd4ed4422be222088dd8535a228b906f43fdc2/modules/public/public.go#L24-L33

guillep2k replied:
I didn't realize that.

guillep2k commented on modules/public/avatars.go:
```diff
@@ -0,0 +1,91 @@
+		}
+	}
+
+	fr, err := fs.NewReader()
```
I think users (admins) should decide which resources are handled locally or remotely. Avatars are too time-sensitive to keep in a remote server, especially if they need to pass though Gitea to be retrieved. Attachments are usually the images of issues and PRs, so they too are pretty much time-sensitive. The `Etag` of course helps a lot. Nonetheless, I'd have three separate settings: one for avatars, one for attachments and one for LFS. Even let the admin set different services for each one.

guillep2k commented on modules/setting/setting.go:
```diff
@@ -774,11 +782,15 @@ func NewContext() {
 
 	InternalToken = loadInternalToken(sec)
 
+	sec = Cfg.Section("storage")
```
As said above, I'd have three sections: `[storage.avatars]`, `[storage.attachments]`, `[storage.lfs]`. Or some flexible system like that.

guillep2k commented on modules/storage/storage.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+// NewReader provides file reader from bucket and error if occurs
+func (fs *FileStorage) NewReader() (*blob.Reader, error) {
```
If you pass an enum here, it's easy to decide between different file storages.

masudur-rahman replied:
If we want to separate storage system for avatars, lfs, attachments ( `[storage.avatars]`, `[storage.attachments]`, `[storage.lfs]` ), we  can use an extra field `BucketURL` here which will be useful to decide between file storages. 
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/modules/storage/storage.go#L22-L26

guillep2k commented:
One important aspect that I didn't see covered: how does one user migrate from one storage system to another? For example: I decide I need aws for my files, but my repository is already up and running. If I change `app.ini`, all my files will be suddenly gone! I'd need some tool to transfer all my files to the new storage and _then_ start Gitea.

masudur-rahman commented on models/migrations/v61.go:
```diff
@@ -29,14 +30,20 @@ func addSizeToAttachment(x *xorm.Engine) error {
+		Path: setting.AttachmentPath,
+	}
+
 	for _, attach := range attachments {
```
Yes, that's very necessary.

guillep2k commented:
A migration tool will be needed indeed. Between different cloud storages I don't know, but at least between any cloud storage and file://. And it should be reversible (e.g. file names should not be problem, even if e.g. avatars get a different name when migrated).

lunny commented:
Should we provide a new command to copy local directory data to a storage?

guillep2k commented:
> 
> 
> Should we provide a new command to copy local directory data to a storage?

@lunny You mean something like this?:

> 
> 
> A migration tool will be needed indeed. Between different cloud storages I don't know, but at least between any cloud storage and file://. And it should be reversible (e.g. file names should not be problem, even if e.g. avatars get a different name when migrated).

lunny commented:
I would like we have an abstract filesystem layer before this PR.  Something like https://godoc.org/github.com/blang/vfs or other better libraries. So we could support both local file system, blob storages and other storage facilities if there are implements for them.

lunny commented:
Or https://github.com/spf13/afero

patcito commented:
Any news about this? @masudur-rahman
